resilience4j-retry/src/main/java/io/github/resilience4j/retry/internal/RetryBudgetImpl.java

```diff
@@ -0,0 +1,208 @@
+            int currentNumOfAttempts = numOfAttempts.get();
+            ringBitSet.setNextBit(false);
+            if (currentNumOfAttempts > 0) {
+                succeededAfterRetryCounter.increment();
+                Throwable throwable = Option.of(lastException.get()).getOrElse(lastRuntimeException.get());
+                publishRetryEvent(() -> new RetryOnSuccessEvent(getName(), currentNumOfAttempts, throwable));
+            } else {
+                succeededWithoutRetryCounter.increment();
+            }
+        }
+
+        public void onError(Exception exception) throws Throwable {
+            if (exceptionPredicate.test(exception)) {
+                lastException.set(exception);
+                throwOrSleepAfterException();
+            } else {
+                failedWithoutRetryCounter.increment();
+                publishRetryEvent(() -> new RetryOnIgnoredErrorEvent(getName(), exception));
+                throw exception;
+            }
+        }
+
+        public void onRuntimeError(RuntimeException runtimeException) {
+            if (exceptionPredicate.test(runtimeException)) {
+                lastRuntimeException.set(runtimeException);
+                throwOrSleepAfterRuntimeException();
+            } else {
+                failedWithoutRetryCounter.increment();
+                publishRetryEvent(() -> new RetryOnIgnoredErrorEvent(getName(), runtimeException));
+                throw runtimeException;
+            }
+        }
+
+        private void throwOrSleepAfterException() throws Exception {
+            int currentNumOfAttempts = numOfAttempts.incrementAndGet();
+            if (canRetry(currentNumOfAttempts)) {
+                waitIntervalAfterFailure();
+            } else {
+                failedAfterRetryCounter.increment();
+                Exception throwable = lastException.get();
+                publishRetryEvent(() -> new RetryOnErrorEvent(getName(), currentNumOfAttempts, throwable));
+                throw throwable;
+            }
+        }
+
+        private void throwOrSleepAfterRuntimeException() {
+            int currentNumOfAttempts = numOfAttempts.incrementAndGet();
+            if (canRetry(currentNumOfAttempts)) {
+                waitIntervalAfterFailure();
+            } else {
+                failedAfterRetryCounter.increment();
+                RuntimeException throwable = lastRuntimeException.get();
+                publishRetryEvent(() -> new RetryOnErrorEvent(getName(), currentNumOfAttempts, throwable));
+                throw throwable;
+            }
+        }
+
+        private void waitIntervalAfterFailure() {
+            // wait interval until the next attempt should start
+            long interval = intervalFunction.apply(numOfAttempts.get());
+            Try.run(() -> sleepFunction.accept(interval))
+                    .getOrElseThrow(ex -> lastRuntimeException.get());
+        }
+
+        private boolean canRetry(final int currentNumOfAttempts) {
```

Comment:
Could you please add some JavaDocs for this `canRetry` method

Reply:
I will add